pkg/filters/oidcadaptor/oidcadaptor.go

```diff
@@ -0,0 +1,428 @@
+			TokenEndpoint:         o.spec.TokenEndpoint,
+			UserinfoEndpoint:      o.spec.UserinfoEndpoint,
+		}
+	}
+	parsed, err := url.Parse(o.spec.RedirectURI)
+	if err != nil {
+		logger.Errorf("parse redirectURI error: %s", err)
+	}
+	o.spec.redirectPath = parsed.Path
+}
+
+func (o *Filter) Inherit(previousGeneration filters.Filter) {
+	o.Init()
+	previousGeneration.Close()
+}
+
+func (o *Filter) Handle(ctx *context.Context) (result string) {
+	req := ctx.GetInputRequest().(*httpprot.Request)
+	var rw *httpprot.Response
+	if rw, _ = ctx.GetOutputResponse().(*httpprot.Response); rw == nil {
+		rw, _ = httpprot.NewResponse(nil)
+		ctx.SetOutputResponse(rw)
+	}
+	spec := o.spec
+	oidcConf := spec.oidcConfig
+	cookieExists := func() bool {
+		for _, c := range req.Cookies() {
+			if spec.CookieName == c.Name {
+				return true
+			}
+		}
+		return false
+	}
+	bearerTokenExists := func() bool {
+		var prefix = "Bearer"
+		authHeader := req.HTTPHeader().Get("Authorization")
+		if strings.HasPrefix(authHeader, prefix) && authHeader[len(prefix):] != "" {
+			return true
+		}
+		return false
+	}
+	if len(spec.CookieName) != 0 && cookieExists() {
+		return ""
+	}
+	if bearerTokenExists() {
+		return ""
+	}
+	if req.Path() == o.spec.redirectPath {
+		return o.handleOidcCallback(ctx)
+	}
+	//https://openid.net/specs/openid-connect-core-1_0.html#AuthorizationEndpoint
+	var authUrlB strings.Builder
+	authUrlB.WriteString(oidcConf.AuthorizationEndpoint)
+	authUrlB.WriteString("?client_id=" + o.spec.ClientId)
+	state := strings.ReplaceAll(uuid.New().String(), "-", "")
+	//state is recommended
+	authUrlB.WriteString("&state=" + state)
+	//user may spend some time doing login stuff
+	err := o.store.put(clusterCacheKey("state", state), "1", 10*time.Minute)
+	if err != nil {
+		logger.Errorf("put oidc state error: %s", err)
+	}
+	reqUrl := func() string {
+		args := req.URL().RawQuery
+		if args != "" {
+			return fmt.Sprintf("%s://%s%s?%s", req.Scheme(), req.Host(), req.Path(), args)
+		}
+		return fmt.Sprintf("%s://%s%s", req.Scheme(), req.Host(), req.Path())
+	}()
+	err = o.store.put(clusterCacheKey("reqUrl", state), reqUrl, 10*time.Minute)
+	if err != nil {
+		logger.Errorf("put origin req url error: %s", err)
+	}
+	//nonce is optional
+	nonce := strings.ReplaceAll(uuid.New().String(), "-", "")
+	authUrlB.WriteString("&nonce=" + nonce)
+	authUrlB.WriteString("&response_type=code")
+	authUrlB.WriteString("&scope=")
+	if len(oidcConf.ScopesSupported) > 0 {
+		authUrlB.WriteString(strings.Join(oidcConf.ScopesSupported, "+"))
+	} else {
+		//authUrlB.WriteString("openid")
+		authUrlB.WriteString("user")
+	}
+	authUrlB.WriteString("&redirect_uri=" + url.QueryEscape(o.spec.RedirectURI))
+	rw.SetStatusCode(http.StatusFound)
+	rw.Header().Set("Location", authUrlB.String())
+	return resultFiltered
+}
+
+func (o *Filter) Status() interface{} {
+	return nil
+}
+
+func (o *Filter) Close() {
+}
+
+func (o *Filter) initDiscoveryOidcConf() {
```

Comment:
per Go naming convention, and there are other cases.

```suggestion
func (o *Filter) initDiscoveryOIDCConf() {
```